applications/OptimizationApplication/python_scripts/optimization_analysis.py

```diff
@@ -0,0 +1,153 @@
+import KratosMultiphysics as Kratos
+from KratosMultiphysics.analysis_stage import AnalysisStage
+from KratosMultiphysics.OptimizationApplication.algorithms.algorithm import Algorithm
+from KratosMultiphysics.OptimizationApplication.model_part_controllers.model_part_controller import ModelPartController
+from KratosMultiphysics.OptimizationApplication.utilities.execution_policy_decorator import ExecutionPolicyDecorator
+from KratosMultiphysics.OptimizationApplication.responses.response_function import ResponseFunction
+from KratosMultiphysics.OptimizationApplication.utilities.control_transformation_technique import ControlTransformationTechnique
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import OptimizationProcessFactory
+
+class OptimizationAnalysis(AnalysisStage):
+    def __init__(self, model: Kratos.Model, project_parameters: Kratos.Parameters):
+        default_parametetrs = Kratos.Parameters("""{
+            "model_parts"                       : [],
+            "analyses"                          : [],
+            "responses"                         : [],
+            "controls"                          : [],
+            "algorithm_settings"                : {},
+            "optimization_info_output_processes": []
+        }""")
+
+        self.model = model
+        self.project_parameters = project_parameters
+        self.project_parameters.AddMissingParameters(default_parametetrs)
+
+        self.optimization_info = OptimizationInfo(self.project_parameters["problem_data"]["echo_level"].GetInt())
+
+        self._CreateModelPartControllers()
+        self._CreateAnalyses()
+        self._CreateResponses()
+        self._CreateControlTechniques()
+
+        super().__init__(model, project_parameters)
+
+    def _GetSimulationName(self):
+        return "OptimizationAnalysis"
+
+    def InitializeSolutionStep(self):
+        # this resets all the saved data for the current time step
+        # so it indicates all the objects using the optimization info
+        # now should recalculate current time step data.
+        self.optimization_info.SetValue("problem_data", {}, 0, True)
+
+        super().InitializeSolutionStep()
+
+    def PrintAnalysisStageProgressInformation(self):
+        Kratos.Logger.PrintInfo(self._GetSimulationName(), "STEP: ", self.optimization_info["step"])
+
+    def KeepAdvancingSolutionLoop(self):
```

Comment:
Move this to algorithm base class . Then change the `Run` method for algorithm based one.